products/data_warehouse/backend/hogql_fixer_ai.py

```diff
@@ -0,0 +1,174 @@
+import os
+from ee.hogai.tool import MaxTool
+from posthog.hogql.database.database import create_hogql_database, serialize_database
+from posthog.hogql.context import HogQLContext
+from ee.hogai.graph.sql.toolkit import SQL_SCHEMA
+from langchain_core.prompts import ChatPromptTemplate
+from langchain_openai import ChatOpenAI
+
+from ee.hogai.graph.schema_generator.parsers import PydanticOutputParserException, parse_pydantic_structured_output
+from ee.hogai.graph.schema_generator.utils import SchemaGeneratorOutput
+
+from posthog.hogql.errors import ExposedHogQLError, ResolutionError
+from posthog.hogql.parser import parse_select
+from posthog.hogql.printer import print_ast
+
+
+_parser_contents: str | None = None
+_lexer_contents: str | None = None
+
+
+def get_parser_contents() -> str:
+    global _parser_contents
+
+    if _parser_contents is not None:
+        return _parser_contents
+
+    with open(f"{os.getcwd()}/posthog/hogql/grammar/HogQLParser.g4") as parser_file:
+        _parser_contents = parser_file.read()
+
+    return _parser_contents
+
+
+def get_lexer_contents() -> str:
+    global _lexer_contents
+
+    if _lexer_contents is not None:
+        return _lexer_contents
+
+    with open(f"{os.getcwd()}/posthog/hogql/grammar/HogQLLexer.g4") as parser_file:
+        _lexer_contents = parser_file.read()
+
+    return _lexer_contents
+
+
+SQL_ASSISTANT_ROOT_SYSTEM_PROMPT = """
+The user has written a HogQL query which contains error. They expect your help with tweaking the HogQL to fix these errors.
+
+IMPORTANT: This is currently your primary task. Therefore `fix_hogql_query` is currently your primary tool.
+Use `fix_hogql_query` when fixing any errors remotely related to HogQL.
+It's very important to disregard other tools for these purposes - the user expects `fix_hogql_query`.
+
+NOTE: When calling the `fix_hogql_query` tool, do not provide any response other than the tool call.
+"""
+
+SYSTEM_PROMPT = f"""
+HogQL is PostHog's variant of SQL. HogQL is a transpiler that outputs Clickhouse SQL. We use Antlr4 to define the HogQL language.
+Below is the antlr parser and lexer definitions - when writing HogQL, ensure you follow the grammar rules.
+
+<hogql_parser>
+{get_parser_contents()}
```

Comment:
I guess because this is being called in a module-level attribute, this will be executed on import. I don't know how long `HogQLLexer.g4` takes to read but it could slow down app startup slightly, and perhaps slow down tests. Might just be worth double checking, or perhaps read this file the first time it is used?

Reply:
I think this will be a nominal amount of extra start-up time. The files being read are pretty small and are on the local filesystem